opm/test_util/summaryIntegrationTest.cpp

```diff
@@ -0,0 +1,352 @@
+	    if(volume.error == 0){
+		std::cout << "For keyword " << keyword << " the total production volume is 0" << std::endl; 
+	    }
+	    else{
+		std::cout << "For keyword " << keyword << " the total production volume is "<< volume.total << ", the error volume is " << volume.error << " the error ratio is " << volume.error/volume.total << std::endl;
+	    }
+	}  
+  	checkForKeyword(timeVec1, timeVec2, keyword);
+	return;
+    } 
+    OPM_THROW(std::invalid_argument, "The keyword used is not common for the two files.");
+  
+}
+
+
+void IntegrationTest::checkForKeyword(std::vector<double> timeVec1, std::vector<double> timeVec2, const char* keyword){
+    std::vector<double> dataVec1, dataVec2;
+  
+    getDataVecs(dataVec1,dataVec2,keyword); 
+    chooseReference(timeVec1, timeVec2,dataVec1,dataVec2);
+ 
+    if(allowSpikes){
+	checkWithSpikes(keyword);
+    }  
+
+    if(findVolumeError ||oneOfTheMainVariables ){
+	volumeErrorCheck(keyword);
+    }
+}
+
+
+
+
+
+int IntegrationTest::checkDeviation(Deviation deviation){
+    double absTol = getAbsTolerance();
+    double relTol = getRelTolerance();
+
+    if (deviation.rel> relTol && deviation.abs > absTol){
+	return 1;
+    }
+    return 0;
+}
+
+
+void IntegrationTest::findGreatestErrorRatio(WellProductionVolume volume, double &greatestRatio, const char* currentKeyword, std::string &keywordWithGreatestErrorRatio){
+    if (volume.total != 0 && volume.total - volume.error > getAbsTolerance()){
+	if(volume.error/volume.total > greatestRatio){
+	    greatestRatio = volume.error/volume.total; 
+	    std::string currentKeywordStr(currentKeyword);
+	    keywordWithGreatestErrorRatio = currentKeywordStr; 
+	}
+    }
+}
+
+
+void IntegrationTest::volumeErrorCheck(const char* keyword){
+    std::string keywordString(keyword);
+    std::string firstFour = keywordString.substr(0,4);
+    std::string fifth = keywordString.substr(4,1); //To make sure we do not include history vectors.
+   
+    if (mainVariable.empty()){
+	updateVolumeError(keyword);
+    }
+    
+    if(firstFour ==  mainVariable && fifth!="H"){
+	if(mainVariable == "WOPR"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WOPR.total += result.total; 
+	    WOPR.error += result.error; 
+	}
+	if(mainVariable == "WWPR"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WWPR.total += result.total; 
+	    WWPR.error += result.error; 
+	}
+
+	if(mainVariable == "WGPR"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WGPR.total += result.total; 
+	    WGPR.error += result.error; 
+	}
+
+	if(mainVariable == "WBHP"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WBHP.total += result.total; 
+	    WBHP.error += result.error; 
+	}	
+    }
+}
+
+
+void IntegrationTest::updateVolumeError(const char* keyword){
+    std::string keywordString(keyword);
```

Comment:
Observe that iff you insist on determing total volumes by integrating rates the correct representation of the function R (t) is as a piecewise constant function,  where the constant value is given as the rate at the upper limit of the interval.

Again - you can query ert with the is_rate property to check the correct temporal interpolation behavior.

If the WEFAC/ GEFAC keywords are different from 1.0 the integration based solution will fail by construction (opm does not support these keywords and maybe never will - so this is maybe not so relevant).